Add SCSS Plugins

rayandrew opened the pull request:
### Added:

1. Razzle SCSS Plugin
2. SCSS Examples
3. Some unit tests

Add little changes to makeLoaderFinder in razzle-dev-utils.
makeLoaderFinder cannot find loader through object like this:

```js
{
  test: /\.(sa|sc)ss$/,
  use: [
    require.resolve('style-loader'),
	require.resolve('css-loader'),
    ...
  ],
}
```

```js
const styleLoaderFinder = makeLoaderFinder('style-loader');
config.module.rules.find(styleLoaderFinder); // will return undefined
```

### Not yet implemented:

- ~~Custom options in SCSS plugins~~

rayandrew committed:
add scss plugin and fixes to makeLoaderFinder

rayandrew committed:
add scss plugin and fixes to makeLoaderFinder

rayandrew committed:
Merge remote-tracking branch 'upstream/master'

rayandrew committed:
add readme for scss plugin

rayandrew committed:
add readme for scss plugin

rayandrew commented:
referencing from issue #662

rayandrew committed:
fix test regex for including .css

rayandrew committed:
fix test regex for including .css

jaredpalmer commented:
Looks good. But tests didn’t actually run.

rayandrew commented:
hmm, i think there's something missing. I cannot reproduce it on my own

rayandrew committed:
remove jest forceExit

rayandrew committed:
remove jest forceExit

rayandrew commented:
@jaredpalmer, is there any special command to run test? 

I have already include test command in scss plugin's package.json

rayandrew committed:
add contributors

rayandrew committed:
add contributors

gregmartyn committed:
preserve the full --inspect= flag as passed to yarn start (#660)

* preserve the full --inspect= flag as passed to yarn start instead of treating it as a boolean

* update --inspect docs & comments

jaredpalmer commented on packages/razzle-plugin-scss/.eslintrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "env": {
+    "jest": true,
+    "node": true
```
Do we need this file in the example?

rayandrew replied:
That's the plugin and I think we need it because eslint screams if i don't put it on eslintrc

jaredpalmer commented on packages/razzle-plugin-scss/README.md:
```diff
@@ -1,3 +1,19 @@
 # razzle-plugin-scss
 
-This is just a stub. Not yet implemented.
+This package contains a plugin for using SCSS with Razzle
```
We should probably document accepted filename endings (.scss, .sass)?

rayandrew replied:
Ah I forgot to put it too. I think it would be nice to put extension in document

jaredpalmer commented on packages/razzle-plugin-scss/README.md:
````diff
@@ -1,3 +1,19 @@
+## Usage in Razzle Projects
+
+```bash
+yarn add razzle-plugin-scss
````
`--dev`

jaredpalmer commented on packages/razzle-plugin-scss/helpers.js:
```diff
@@ -0,0 +1,16 @@
+  postCssLoaderFinder,
+  resolveUrlLoaderFinder,
+  sassLoaderFinder,
+  styleLoaderFinder,
```
Should we document in README what loaderFinders are being exported so people can reuse them?

rayandrew replied:
It would be a good idea! I don't know whether some loaders are reusable at first

jaredpalmer commented on packages/razzle-plugin-scss/package.json:
```diff
@@ -1,6 +1,24 @@
+    "sass-loader": "^7.0.3"
+  },
+  "devDependencies": {
+    "jest": "20.0.4"
```
Bump to jest 23

rayandrew replied:
Will upgrade it. But may I ask a question?

```js
"husky": "^0.14.3",
"jest": "^22.4.3",
"lerna": "2.10.1",
```

razzle lerna package.json still using jest 22.4.3, should I bump it to 23 or stay in 22.4.3?

jaredpalmer commented on packages/razzle-plugin-scss/package.json:
```diff
@@ -1,6 +1,24 @@
+    "mini-css-extract-plugin": "^0.4.0",
+    "node-sass": "^4.9.0",
+    "postcss-flexbugs-fixes": "^3.3.1",
+    "razzle": "^2.2.0",
```
Razzle is not a dep.

rayandrew replied:
Ah I see, can I make it as dev dependency like typescript-plugin?

jaredpalmer commented on packages/razzle-plugin-scss/package.json:
```diff
@@ -1,6 +1,24 @@
+  "dependencies": {
+    "autoprefixer": "^8.6.2",
+    "mini-css-extract-plugin": "^0.4.0",
+    "node-sass": "^4.9.0",
```
Use `node-sass-chokidar`

jaredpalmer replied:
it's much less flaky.

rayandrew replied:
I want too, but [sass-loader](https://github.com/webpack-contrib/sass-loader) make it as peer dependencies and using node-sass API.

Are you perhaps successfully combine webpack and node-sass-chokidar together?

jaredpalmer replied:
Yeah it's just a warning. It works fine. We've used it for almost a year.

rayandrew replied:
Gotcha!

jaredpalmer commented on packages/razzle-plugin-scss/tests/index.test.js:
```diff
@@ -0,0 +1,113 @@
+
+      it('should add style-loader', () => {
+        const rule = config.module.rules.find(styleLoaderFinder);
+        expect(rule).not.toBeUndefined();
```
When I have a lot of tests that have exactly the same structure, I prefer to just keep them in an array and just call forEach on them. Helps stay DRY.  

For example:

```js
const devLoaderTests = [{
   name: '...',
   loaderFinder: '...'
}, {
 // ...
},
 {
 // ...
}]


describe('dev', () => {
  devLoaderTests.forEach(item => {
    it(item.name, () => {
     expect(config.module.rules.find(item.loaderFinder).not.toBeUndefined());
    })
  });
});
```

rayandrew replied:
Ahh sorry for repeating codes. I will fix it!

rayandrew replied:
should I make it recursive? 

some tests depends on environment like:

```js
describe('razzle-scss-plugin', () => {
  describe('when creating a web config', () => {
    describe('when environment set to development', () => {

	});

	describe('when environment set to production', () => {

	});
  );

  describe('when creating a node config', () => {
    describe('when environment set to development', () => {

	});

	describe('when environment set to production', () => {

	});
  );
);
```

or better if I do like this:

```js
describe('razzle-scss-plugin', () => {
  describe('web config on development', () => {

  });

  describe('web config on production', () => {

  });

  describe('node config on development', () => {

  });

  describe('node config on production', () => {

  });
);
```

what do you think?

rayandrew replied:
here's what I do in mocha + chai for recursive **(will change to suitable tests for razzle)**:

```js
  const template = (strings, ...keys) => (...values) => {
    const dict = values[values.length - 1] || {};
    const result = [strings[0]];
    keys.forEach((key, i) => {
      const value = Number.isInteger(key) ? values[key] : dict[key];
      result.push(value, strings[i + 1]);
    });
    return result
      .join("")
      .replace(/\s+/g, " ")
      .trim();
  },

  const generateTest = (authenticated, credential, templates, tests) => {
    return _.zipWith(templates, tests, (template, test) => {
      if (_.isFunction(template)) {
        it(template({ ...test.props, ...credential }), async () => {
          if (_.isObject(test)) {
            if (_.isUndefined(test.payload)) {
              test.check(
                await authenticated[test.method](
                  module.exports.baseUrl(test.url)
                )
              );
            } else {
              test.check(
                await authenticated[test.method](
                  module.exports.baseUrl(test.url)
                ).send(test.payload)
              );
            }
          }
        });
      } else {
        describe(template.desc({ role: credential.role }), () => {
          return module.exports.generateTest(
            authenticated,
            credential,
            template.test,
            test
          );
        });
      }
    });
```

```js
const templateTest = [
  template`should ${"should"} get list of users if ${"role"} is logged in`,
  template`should ${"should"} edit users if ${"role"} is logged in`,
  {
    desc: template`details of specific user by ${"role"}`,
    test: [
      template`should ${"should"} get details of specific wrong id of user because resource not found`,
      template`should ${"should"} get details of specific correct id of user`
    ]
  }
];
```

```js
const authenticatedTest = [{
   credentials: {
    username: "administrator",
    password: "administrator",
    role: "admin"
   },
   tests: [{
     url: "/users",
     method: `get`,
     check: result => {
      expect(result).to.have.status(200);
      expect(result.body).to.have.ownProperty("data");
      expect(result.body.data).to.be.an("array");
      expect(result.body).to.have.all.keys(paginateProps);
     }
    }, {
     props: {
      should: "not"
     },
     url: `/users/${user.id_user}`,
     payload: user,
     method: `patch`,
     check: result => {
      expect(result).to.have.status(403);
      expect(result.body.message).to.equal("Forbidden");
      expect(result.body.name).to.equal("ForbiddenError");
     }
    },
    [{
     props: {
      should: "not"
     },
     url: `/user/${user.wrong_id_user}`,
     method: `get`,
     check: result => {
      expect(result).to.have.status(404);
      expect(result.body.message).to.equal("User not found.");
      expect(result.body.name).to.equal("NotFoundError");
     }
    }, {
     url: `/user/${user.id_user}`,
     method: `get`,
     check: result => {
      expect(result).to.have.status(200);
      expect(result.body).to.be.an("object");
      expect(result.body).to.include.all.keys(userColumns);
     }
    }],
    ...
   ];
```

and then generate tests like this

```js
 authenticatedTest.forEach(auth => {
   describe(`authenticated role : ${auth.credentials.role}`, () => {
    const authenticated = chai.request.agent(routes);

    before(() =>
     knex.migrate
     .rollback()
     .then(() => knex.migrate.latest())
     .then(() => knex.seed.run())
     .then(() =>
      authenticated
      .post(`${config.get("routePrefix")}/session`)
      .send(auth.credentials)
     )
     .then(res => {
      expect(res).to.have.status(200);
      expect(res.body).to.have.ownProperty("username");
      expect(res.body).to.have.ownProperty("status");
     })
    );

    after(() =>
     authenticated.del(`${config.get("routePrefix")}/session`).then(res => {
      expect(res).to.have.status(200);
      expect(res.body).to.be.an("object");
      expect(res.body.message).to.equal("Logged out successfully.");
     })
    );

    generateTest(authenticated, auth.credentials, templateTest, auth.tests);
  });
```

what do you think?

jaredpalmer commented on packages/razzle-plugin-scss/tests/index.test.js:
```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+const createConfig = require('razzle/config/createConfig');
```
Ignore prior comment. Make razzle a `devDep`.

jaredpalmer commented on packages/razzle-plugin-scss/tests/index.test.js:
```diff
@@ -0,0 +1,113 @@
+      expect(rule).toBeUndefined();
+    });
+  });
+});
```
Remove yarn.lock file please. It will mess up yarn module hoisting

jaredpalmer reviewed (changes requested):
This is truly awesome! You rock! Just a few nits here and there.

jaredpalmer commented:
So I think our `less` and `postcss` plugins are going to share A LOT of code with this. Do we think it makes sense to abstract some of this into a `razzle-plugin-style` package and then use that the core for each one? That is, the `less`, `sass`, `postcss` plugins would really just be sort of shell packages that pass in some configs, while the workhorse would be the `razzle-plugin-style` or like `razzle-style-utils` or whatever we want to call it.

davidnguyen11 committed:
added example razzle + react router 3 (#539)

* added example razzle + react router 3

* Remove yarn.lock

* Fix pkg.json

Commit:
Add polka example (#520)

* Add polka example

Serves as a generic express middleware-compatible server example. (related to jaredpalmer/after.js#77)

* Remove with-polka/setupTests.js

* Remove yarn.lock

rayandrew commented:
That will be a great idea! It seems postcss is used too many times for sure. 
It will be great to make it reusable for other plugins too.

PS: gonna make some commits tomorrow with the fixes

rayandrew committed:
refactor tests

rayandrew committed:
refactor tests

rayandrew committed:
add plugin options

rayandrew committed:
add plugin options

rayandrew committed:
edit readme for better options knowledge

rayandrew committed:
edit readme for better options knowledge

rayandrew commented:
any thoughts about this @jaredpalmer?

jaredpalmer commented:
Sorry was at an offsite yesterday. Looking right meow 🐱

jaredpalmer commented on examples/with-scss/src/App.scss:
```diff
@@ -1,5 +1,7 @@
+@import url('https://fonts.googleapis.com/css?family=Open+Sans');
```
cool!

jaredpalmer commented on packages/razzle-plugin-scss/README.md:
```diff
@@ -1,11 +1,11 @@
 # razzle-plugin-scss
 
-This package contains a plugin for using SCSS with Razzle
+This package contains a plugin for using SCSS/SASS with Razzle
```
Link to the sass website here?

rayandrew replied:
gotcha!

jaredpalmer commented on packages/razzle-plugin-scss/README.md:
````diff
@@ -16,4 +16,185 @@ Using the plugin with the default options
+
+```js
+// default options
+const defaultOptions = {
````
When i show off options. Show an example of how to use with a simple change and shortest possible snippet. Then below show/describe defaults beneath. It's just overwhelming here.

rayandrew replied:
Better if I just delete it? I have already shown the default options beneath the example

jaredpalmer commented on packages/razzle-plugin-scss/README.md:
```diff
@@ -16,4 +16,185 @@ Using the plugin with the default options
+      ident: 'postcss',
+    },
+    plugins: [
+      PostCssFlexBugFixes,
```
Not sure I'm in love with this API. Let me think about this for a few hours and come back with some ideas.

jaredpalmer reviewed (changes requested):
Almost there. Going to discuss the configuration API with my team. Will ping with updates shortly

rayandrew committed:
change node-sass to node-sass-chokidar

rayandrew committed:
change node-sass to node-sass-chokidar

bdo committed:
Adding testEnvironmentOptions to supportedKeys for Jest (#678)

rayandrew commented:
any updates or changes @jaredpalmer?

ambar committed:
fix koa hmr (#693)

jackjocross committed:
Add pretty node errors (#691)

jackjocross committed:
Ignore column from source map and only use pretty errors if file is read (#696)

alexparish committed:
Add documentation for CSS Modules (#694)

kireerik committed:
Erik Engi added to the contributors section (#700)

rayandrew commented:
any updates @jaredpalmer?

jaredpalmer committed:
Update README.md

ericvicenti committed:
Introduce config.modifyBabelOptions (#702)

I’ve found that the babel configuration is important to modify in my environment.

Specifically in my env, I’d like to set `babelrc: false` so that the .babelrc file is ignored, and a custom babel configuration could be provided in the `razzle.config.js`.

ambar committed:
add support for function as plugin (#704)

jeetiss committed:
add mdx loader (#711)

jeetiss committed:
Mdx improve (#712)

* add peer dependency, ipdate install instruction

* update readme

ericclemmons committed:
Fix .mjs issue with GraphQL 0.13.2 (#715)

* Fix .mjs issue with GraphQL 0.13.2

https://github.com/graphql/graphql-js/issues/1272#issuecomment-377384574

* Fix .mjs still referencing "require"

https://github.com/aws-amplify/amplify-js/issues/686#issuecomment-387710340

jaredpalmer committed:
remove package-lock.json

jaredpalmer committed:
Make readme's match

jaredpalmer committed:
v2.3.0

jaredpalmer committed:
Revert "v2.3.0"

This reverts commit 361bb0c9267cf543f775207ca4b296ba18d29ba3.

jaredpalmer committed:
v2.4.0

jpetitcolas committed:
Fix production sourcemaps (#717)

jaredpalmer commented:
Reviewing now. Thanks for the ping

jaredpalmer commented:
If we can fix conflicts will merge.

rayandrew commented:
i think i need to rebase from the new master branch. will soon resolve the conflicts

thank you for your response @jaredpalmer!

rayandrew committed:
bumped package version and fix conflicts

rayandrew committed:
bumped package version and fix conflicts

rayandrew committed:
bumped package version and fix conflicts

rayandrew committed:
edited readme

rayandrew commented:
resolved the conflict, wanna hear your thoughts @jaredpalmer

ion-willo commented:
any updates @jaredpalmer?

mzedeler commented:
I could use this plugin too.

zhangking commented:
When will the plan be released?

snaerth commented:
Any news on the release

willian commented:
I am interested in using this too.

nealoke commented:
+1

snaerth commented:
No news yet. @rayandrews  would it be possible for you to provide a razzle.config.js for scss and css modules?

jaredpalmer commented:
@crosscompile can you review on Monday?

jackjocross commented:
The PR changes LGTM!  I agree this will probably share some code with `css` and `postcss` plugins but we can probably add an abstraction around those later.

The history in this PR got a little crazy so I pulled out the relevant changes into #786, @rayandrews let me know if anything is missing from that PR.

rayandrew commented:
Looks good to me too!
Maybe we can move discussion for this feature in PR #786 and close this one?

cc: @jaredpalmer @crosscompile
